services/codecov/codecov.service.js

```diff
@@ -1,108 +1,139 @@
 'use strict'
 
-const queryString = require('query-string')
-const LegacyService = require('../legacy-service')
-const { makeBadgeData: getBadgeData } = require('../../lib/badge-data')
-const {
-  coveragePercentage: coveragePercentageColor,
-} = require('../color-formatters')
+const Joi = require('joi')
+const { coveragePercentage } = require('../color-formatters')
+const { BaseJsonService } = require('..')
 
-// This legacy service should be rewritten to use e.g. BaseJsonService.
-//
-// Tips for rewriting:
-// https://github.com/badges/shields/blob/master/doc/rewriting-services.md
-//
-// Do not base new services on this code.
-module.exports = class Codecov extends LegacyService {
+// https://docs.codecov.io/reference#totals
+// A new repository that's been added but never had any coverage reports
+// uploaded will not have a `commit` object in the response and some times
+// the `totals` object can also be missing for the latest commit.
+// Keeping these as optional in the schema and handling it in the transform
+// function to maintain consistent badge behavior with the legacy service implementation.
+const schema = Joi.object({
+  commit: Joi.object({
+    totals: Joi.object({
+      c: Joi.number().required(),
+    }),
+  }),
+}).required()
+
+const queryParamSchema = Joi.object({
+  token: Joi.string(),
+}).required()
+
+const documentation = `
+  <p>
+    You may specify a Codecov token to get coverage for a private repository.
+  </p>
+  <p>
+    See the <a href="https://docs.codecov.io/reference#authorization">Codecov Docs</a>
+    for more information about creating a token.
```

Comment:
As an aside, do you know if these are read-only / status-only tokens? I hate to remove wanted functionality, though at some point we really should lock down the remaining endpoints that accept read-write tokens. For the purpose of this PR preserving the behavior is 👍

Reply:
I don't know, and it wasn't clear to me from the docs whether there were different permission scopes. My only experience with Codecov tokens has been repository-specific tokens that have permissions to publish coverage results

Reply:
I don't think their anchors are working, but:

https://codecov.io/site/security#how-does-codecov-authenticate-access-to-a-repository
https://codecov.io/site/security#what-scope-is-used-when-signing-up-using-github 

TL;DR if I'm reading correctly (at least for GH), scope carries through to Codecov perms, based on the token-owners permissions on the repo. So a user with read-only access to a private GH repo would create a token that would only be able to be used to see/read Codecov coverage info for that repo. A user with read/write access to that private GH repo would create a token that could both see/read coverage info, and publish/upload coverage results for that repo in Codecov.

Reply:
Sounds like it is _possible_ that folks might create separate account that enable this to be done securely, though it's also possible they won't have…